### PR TITLE
Combine section_init() and create_subdir()

### LINKED ISSUE #2644: `init_site()` doesn't have an `override` argument
Are you sure we shouldn't just call`section_init()` here? I also noticed that `init_site()` lacks `override`, and I needed it when I started experimenting with quarto.

That would make it more consistent with the other functions

_Originally posted by @hadley in https://github.com/r-lib/pkgdown/pull/2642#discussion_r1630237712_

### PR DESCRIPTION
**Pull request by hadley**
* We can figure out `depth` based on whether or not a `subdir` is supplied
* Ensures all user facing functions have the same interface (i.e. have `override`)
* Brings together directory checking code in one place

Fixes #2644

@olivroy I started reviewing #2645 but then I wondered if maybe we could simplify things by unifying `section_init()` and `create_subdir()` and that resulted in this PR 😄

### CODE REVIEW

File: R/context.R [hunk 1]
```diff
@@ -1,13 +1,41 @@
-section_init <- function(pkg, depth, override = list(), .frame = parent.frame()) {
+section_init <- function(pkg,
+                         subdir = NULL,
+                         override = list(),
+                         .frame = parent.frame()) {
+  rstudio_save_all()
```
Comment (hadley) [inline R/context.R:5]: We want to save all files before we read in the yaml.

**Comment by olivroy** [comment]
I think it is a good idea. But this somehow seems to break the workflow I created in #2571.

i.e. to call `build_home()` yields the error message that you need to call `init_site()`.

**Comment by olivroy** [comment]
Is the no-pandoc failure stochastic ? If so, maybe we should skip this test if no pandoc?

**Comment by hadley** [comment]
I figured out a reprex:

```R
src_path <- withr::local_tempdir()
writeLines(path(src_path, "README.md"), "Hi")
description <- desc::desc("!new")
description$set("Package", "testpackage")
description$set("Title", "A test package")
description$write(file = path(src_path, "DESCRIPTION"))    

build_home(src_path)
clean_site(src_path)
build_home(src_path)
```

File: Hi
Comment (olivroy) [inline Hi:1]: Remove this?

**Review by olivroy — APPROVED** [review]
Now works for me, thanks! I like the idea. Will finish up #2642 after this. Maybe add that this closes #2645 

Is this news worthy? probably not :)

File: Hi
Comment (hadley) [reply Hi:1]: Oops

**Comment by hadley** [comment]
IMO this isn't worth its own a news bullet because it's more of a refinement of existing work that already has news bullets.

**Pull request merged**